File: src/lib/pipelineNotes.js
import React from 'react';


export const humanizePipelineDetail = (detail = '') => {
  const text = String(detail || '').replace(/\s+/g, ' ').trim();
  const lowered = text.toLowerCase();
  if (!text) return 'Starting the pipeline.';
  if (lowered.startsWith('auto-enabled')) return 'Search intent detected. Auto-enabling web research and collecting sources.';
  if (lowered.includes('planning up to')) return 'Planning the query first, so the search is not random.';
  if (lowered.includes('round') && lowered.includes('follow-up')) return `Initial results were weak, trying the next search angle: ${text}`;
  if (lowered.includes('search') && lowered.includes(':')) return `Running search: ${text.split(':').slice(1).join(':').trim() || text}`;
  if (lowered.startsWith('→')) return `Found source: ${text.replace(/^→\s*/, '')}`;
  if (lowered.includes('collected') && lowered.includes('unique sources')) return `Collecting sources: ${text}`;
  if (lowered.includes('semantic retrieval')) return `Searching local files for relevant chunks: ${text}`;
  if (lowered.startsWith('searching')) return `Scanning uploaded files: ${text}`;
  if (lowered.includes('analysis plan ready')) return `Plan is ready. Building the answer against the evidence now.`;
  if (lowered.includes('calling') && lowered.includes('understand')) return `Understanding the question's intent and building the answer structure.`;
  if (lowered.startsWith('preparing')) return `Composing the draft: ${text}`;
  if (lowered.includes('synthesizing')) return `Merging sources and writing the final answer.`;
  if (lowered.includes('verify') || lowered.includes('quality')) return `Checking answer quality and grounding.`;
  if (lowered.includes('repair')) return `Found a gap, refining the answer.`;
  if (lowered.includes('answer ready') || lowered.includes('ready')) return `Answer is ready.`;
  if (lowered.includes('still') || lowered.includes('active')) return `Still working: ${text}`;
  return text.length > 170 ? `${text.slice(0, 167)}...` : text;
};

export const buildWorkingNotes = (events = [], pipeline = {}) => {
  const notes = [];
  const candidates = [
    ...events.filter(event => event.detail).map(event => ({
      id: `${event.at || ''}-${event.stage || ''}-${event.detail}`,
      stage: event.stage || pipeline.stage || 'working',
      text: humanizePipelineDetail(event.detail),
      live: false,
    })),
  ];
  if (pipeline.detail) {
    candidates.push({
      id: `current-${pipeline.stage}-${pipeline.detail}`,
      stage: pipeline.stage || 'working',
      text: humanizePipelineDetail(pipeline.detail),
      live: true,
    });
  }
  const seen = new Set();
  for (const item of candidates.reverse()) {
    const key = item.text.toLowerCase();
    if (!item.text || seen.has(key)) continue;
    seen.add(key);
    notes.unshift(item);
    if (notes.length >= 4) break;
  }
  return notes.length ? notes : [{ id: 'start', stage: 'starting', text: 'Got it. Processing the request.', live: true }];
};

export const directActivityToNote = item => {
  const label = item?.label || '';
  const detail = item?.detail || '';
  if (/sending/i.test(label)) return `Request sent: ${detail}`;
  if (/connecting/i.test(label)) return `Connecting to the model: ${detail}`;
  if (/streaming/i.test(label)) return `Answer is streaming in: ${detail}`;
  if (/saving/i.test(label)) return `Saving chat history.`;
  if (/stopped/i.test(label)) return `Stopped. You can change the model here and ask again.`;
  return detail || label || 'Working...';
};

File: ios/Locus/Locus/Lib/PipelineNotes.swift
import Foundation

/// Live "working notes" for the pipeline card — a 1:1 port of `src/lib/pipelineNotes.js` so the
/// running commentary reads the same on both clients.
enum PipelineNotes {
    struct Note: Identifiable, Equatable {
        let id: String
        let text: String
        let live: Bool
    }

    static func humanize(_ detail: String) -> String {
        let text = detail.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return "Starting the pipeline." }
        let lowered = text.lowercased()

        if lowered.hasPrefix("auto-enabled") {
            return "Search intent detected. Auto-enabling web research and collecting sources."
        }
        if lowered.contains("planning up to") {
            return "Planning the query first, so the search is not random."
        }
        if lowered.contains("round") && lowered.contains("follow-up") {
            return "Initial results were weak, trying the next search angle: \(text)"
        }
        if lowered.contains("search") && lowered.contains(":") {
            let tail = text.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false)
                .dropFirst().joined(separator: ":").trimmingCharacters(in: .whitespaces)
            return "Running search: \(tail.isEmpty ? text : tail)"
        }
        if text.hasPrefix("→") {
            return "Found source: \(text.replacingOccurrences(of: "^→\\s*", with: "", options: .regularExpression))"
        }
        if lowered.contains("collected") && lowered.contains("unique sources") {
            return "Collecting sources: \(text)"
        }
        if lowered.contains("semantic retrieval") {
            return "Searching local files for relevant chunks: \(text)"
        }
        if lowered.hasPrefix("searching") { return "Scanning uploaded files: \(text)" }
        if lowered.contains("analysis plan ready") {
            return "Plan is ready. Building the answer against the evidence now."
        }
        if lowered.contains("calling") && lowered.contains("understand") {
            return "Understanding the question's intent and building the answer structure."
        }
        if lowered.hasPrefix("preparing") { return "Composing the draft: \(text)" }
        if lowered.contains("synthesizing") { return "Merging sources and writing the final answer." }
        if lowered.contains("verify") || lowered.contains("quality") {
            return "Checking answer quality and grounding."
        }
        if lowered.contains("repair") { return "Found a gap, refining the answer." }
        if lowered.contains("answer ready") || lowered.contains("ready") { return "Answer is ready." }
        if lowered.contains("still") || lowered.contains("active") { return "Still working: \(text)" }
        return text.count > 170 ? String(text.prefix(167)) + "..." : text
    }

    /// Newest four distinct notes, oldest first, with the current pipeline detail marked live.
    static func build(events: [[String: AnyCodable]], stage: String, detail: String) -> [Note] {
        var candidates: [Note] = events.compactMap { event in
            guard let eventDetail = event["detail"]?.string, !eventDetail.isEmpty else { return nil }
            let at = event["at"]?.string ?? ""
            let eventStage = event["stage"]?.string ?? stage
            return Note(id: "\(at)-\(eventStage)-\(eventDetail)", text: humanize(eventDetail), live: false)
        }
        if !detail.isEmpty {
            candidates.append(Note(id: "current-\(stage)-\(detail)", text: humanize(detail), live: true))
        }

        var notes: [Note] = []
        var seen = Set<String>()
        for item in candidates.reversed() {
            let key = item.text.lowercased()
            if item.text.isEmpty || seen.contains(key) { continue }
            seen.insert(key)
            notes.insert(item, at: 0)
            if notes.count >= 4 { break }
        }
        return notes.isEmpty
            ? [Note(id: "start", text: "Got it. Processing the request.", live: true)]
            : notes
    }

    /// "1m 4s" / "2h 5m" — the web's `formatElapsedTime`.
    static func elapsed(_ totalSeconds: Int) -> String {
        let days = totalSeconds / 86_400
        let hours = (totalSeconds % 86_400) / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        if days > 0 { return "\(days)d \(hours)h" }
        if hours > 0 { return "\(hours)h \(minutes)m" }
        if minutes > 0 { return "\(minutes)m \(seconds)s" }
        return "\(seconds)s"
    }
}
